components/pluralrules/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+//! [`Plural Category`]: ./enum.PluralCategory.html
+//! [`Language Plural Rules`]: https://unicode.org/reports/tr35/tr35-numbers.html#Language_Plural_Rules
+//! [`CLDR`]: http://cldr.unicode.org/
+//! [`data`]: ./data/index.html
+pub mod data;
+mod operands;
+pub mod rules;
+
+use icu_locale::LanguageIdentifier;
+pub use operands::PluralOperands;
+
+/// A type of a plural rule which can be associated with the [`PluralRules`] struct.
+///
+/// [`PluralRules`]: ./struct.PluralRules.html
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
+pub enum PluralRuleType {
+    /// Cardinal plural forms express units such as time, currency or distance, used in conjunction
+    /// with a number expressed in decimal digits (i.e. "2", not "two").
+    ///
+    /// For example, English has two forms for cardinals:
+    ///
+    /// * [`One`]: `1 day`
+    /// * [`Other`]: `0 days`, `2 days`, `10 days`, `0.3 days`
+    ///
+    /// [`One`]: ./enum.PluralCategory.html#variant.One
+    /// [`Other`]: ./enum.PluralCategory.html#variant.Other
+    Ordinal,
+    /// Ordinal plural forms denote the order of items in a set and are always integers.
+    ///
+    /// For example, English has four forms for ordinals:
+    ///
+    /// * [`One`]: `1st floor`, `21st floor`, `101st floor`
+    /// * [`Two`]: `2nd floor`, `22nd floor`, `102nd floor`
+    /// * [`Few`]: `3rd floor`, `23rd floor`, `103rd floor`
+    /// * [`Other`]: `4th floor`, `11th floor`, `96th floor`
+    ///
+    /// [`One`]: ./enum.PluralCategory.html#variant.One
+    /// [`Two`]: ./enum.PluralCategory.html#variant.Two
+    /// [`Few`]: ./enum.PluralCategory.html#variant.Few
+    /// [`Other`]: ./enum.PluralCategory.html#variant.Other
+    Cardinal,
+}
+
+/// The plural categories are used to format messages with numeric placeholders, expressed as decimal numbers.
+///
+/// The fundamental rule for determining plural categories is the existence of minimal pairs: whenever two different
+/// numbers may require different versions of the same message, then the numbers have different plural categories.
+///
+/// All languages supported by `ICU4X` can match any number to one of the categories.
+///
+/// # Examples
+///
+/// ```
+/// use icu_locale::LanguageIdentifier;
+/// use icu_pluralrules::{PluralRules, PluralRuleType, PluralCategory};
+/// use icu_pluralrules::data::provider::DummyDataProvider;
+///
+/// let lang: LanguageIdentifier = "en".parse()
+///     .expect("Failed to parse a language identifier.");
+///
+/// let dp = DummyDataProvider::default();
+///
+/// let pr = PluralRules::try_new(lang, PluralRuleType::Cardinal, &dp)
+///     .expect("Failed to construct a PluralRules struct.");
+///
+/// assert_eq!(pr.select(1_usize), PluralCategory::One);
+/// assert_eq!(pr.select(5_usize), PluralCategory::Other);
+/// ```
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
+pub enum PluralCategory {
+    /// This category is used for message forms that are specific to cases
```

Comment:
yes, they're just names and in particular they should not be expected to overlap with numbers (e.g. category `One` may match digit `2`). I'll try to improve the docs.

Reply:
Let me know if I succeeded in improving the docs  with the `Note` by `One`.

Reply:
I think it's probably best to err on the side of under-defining the entries in this enum.  The note under ONE helps, but I don't think it solves the confusion.

AFAIK, UTS 35 doesn't require anything specific from any plural form.  It's completely up to the rule strings to decide what each category means.  Any meanings about numerical values are only by under-defined convention.  It isn't our job in ICU4X to write those conventions down.

Reply:
@sffc can you suggest comment example that would be on the right level of detail for you?

Reply:
I'll defer to Shane on this. Based on what he said, my guess about what it could say would be pretty nondescript, like, "Represents the ONE plural category", and you could put a mention to UTS 35 and/or link to it in the docstrings.  Personally, I like the descriptive example comments (ex: giving a sense of what it means in English/Welsh/Polish/etc.), but you could change precise words like "is" to less precise words like "might be" since we won't/can't be precise about meanings.

Reply:
> @sffc can you suggest comment example that would be on the right level of detail for you?

Here goes:

```rust
pub enum PluralCategory {
    /// CLDR "zero" plural category. Used in Arabic and Latvian, among others.
    ///
    /// Examples of numbers having this category:
    ///
    /// - 0 in Arabic (ar), Latvian (lv)
    /// - 10~20, 30, 40, 50, ... in Latvian (lv)
    Zero,
    /// CLDR "one" plural category. Signifies the singular form in many languages.
    ///
    /// Examples of numbers having this category:
    ///
    /// - 0 in French (fr), Portuguese (pt), ...
    /// - 1 in English (en) and most other languages
    /// - 2.1 in Filipino (fil), Croatian (hr), Latvian (lv), Serbian (sr)
    /// - 2, 3, 5, 7, 8, ... in Filipino (fil)
    One,
    /// CLDR "two" plural category. Used in Arabic, Hebrew, and Slovenian, among others.
    ///
    /// Examples of numbers having this category:
    ///
    /// - 2 in Arabic (ar), Hebrew (iw), Slovenian (sl)
    /// - 2.0 in Arabic (ar)
    Two,
    /// CLDR "few" plural category. Used in Romanian, Polish, Russian, and others.
    ///
    /// Examples of numbers having this category:
    ///
    /// - 0 in Romanian (ro)
    /// - 1.2 in Croatian (hr), Romanian (ro), Slovenian (sl), Serbian (sr)
    /// - 2 in Polish (pl), Russian (ru), Czech (cs), ...
    /// - 5 in Arabic (ar), Lithuanian (lt), Romanian (ro)
    Few,
    /// CLDR "many" plural category. Used in Polish, Russian, Ukrainian, and others.
    ///
    /// Examples of numbers having this category:
    ///
    /// - 0 in Polish (pl)
    /// - 1.0 in Czech (cs), Slovak (sk)
    /// - 1.1 in Czech (cs), Lithuanian (lt), Slovak (sk)
    /// - 15 in Arabic (ar), Polish (pl), Russian (ru), Ukrainian (uk)
    Many,
    /// CLDR "other" plural category, used as a catch-all. Each language supports it, and it
    /// is also used as a fail safe result for in case no better match can be identified.
    ///
    /// In some languages, such as Japanese, Chinese, Korean, and Thai, this is the only
    /// plural category.
    ///
    /// Examples of numbers having this category:
    ///
    /// - 0 in English (en), German (de), Spanish (es), ...
    /// - 1 in Japanese (ja), Korean (ko), Chinese (zh), Thai (th), ...
    /// - 2 in English (en), German (de), Spanish (es), ...
    Other,
}
```

Reply:
Thank you! Applied!